In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import warnings 
warnings.filterwarnings('ignore')

pd.set_option('display.max_rows',None)
pd.set_option('display.max_columns',None)
pd.set_option('display.expand_frame_repr',False)
pd.set_option('max_colwidth',None)

In [2]:
ipl_df = pd.read_csv(r"C:\Users\ANAIKUTTY\Documents\Cricket analysis\data files\ipl_2008_2024_ball_by_ball_data.csv")

In [3]:
df = ipl_df.copy()

In [4]:
df['phase'] = df['ball'].apply(lambda x:'powerplay' if x < 6 else ( 'middle' if x < 15 else 'death over'))

In [5]:
df['isdot'] = df['runs_off_bat'].apply(lambda x: 1 if x == 0 else 0)
df['isone'] = df['runs_off_bat'].apply(lambda x: 1 if x == 1 else 0)
df['istwo'] = df['runs_off_bat'].apply(lambda x: 1 if x == 2 else 0)
df['isthree'] = df['runs_off_bat'].apply(lambda x: 1 if x == 3 else 0)
df['isfour'] = df['runs_off_bat'].apply(lambda x: 1 if x == 4 else 0)
df['issix'] = df['runs_off_bat'].apply(lambda x: 1 if x == 6 else 0)

In [12]:
pp_df = df[df['phase']== 'powerplay']

In [17]:
pp_df.head(1)

striker  matches  runs  balls  outs  four  six  dots  strikerate  average  boundary_%      dot_%
0  A Ashish Reddy        1     5      7     1     1    0     5   71.428571      5.0         4.0  71.428571

In [13]:
pp_df = pd.DataFrame(pp_df.groupby('striker').agg({'match_id':'nunique','runs_off_bat':'sum','ball':'count','player_dismissed':'count','isfour':'sum','issix':'sum','isdot':'sum'})).reset_index()
pp_df = pp_df.rename(columns={'match_id':'matches','runs_off_bat':'runs','ball':'balls','player_dismissed':'outs','isfour':'four','issix':'six','isdot':'dots'})

In [24]:
pp_df['strikerate'] = pp_df['runs'] / pp_df['balls'] * 100
pp_df['average'] = np.where(pp_df['outs']==0,pp_df['runs'],pp_df['runs'] / pp_df['outs'])
pp_df['boundary_%']  = ((pp_df['four']*4 + pp_df['six']* 6) / pp_df['runs']) *100
pp_df['dot_%'] = pp_df['dots'] / pp_df['balls'] *100

In [25]:
pp_df = pp_df[pp_df['matches'] >= 20].sort_values(by='strikerate',ascending=False)

In [ ]:
pp_df.head(15)
#matches played = 20

striker  matches  runs  balls  outs  four  six  dots  strikerate    average  boundary_%      dot_%
331          TM Head       20   412    211     8    53   22    77  195.260664  51.500000   83.495146  36.492891
298        SP Narine       58   946    577    41   119   55   288  163.951473  23.073171   85.200846  49.913345
40   Abhishek Sharma       40   758    501    22    90   35   218  151.297405  34.454545   75.197889  43.512974
354      YBK Jaiswal       52  1009    677    25   149   36   328  149.039882  40.360000   80.475719  48.449040
133      JM Bairstow       47   864    606    22   113   34   279  142.574257  39.272727   75.925926  46.039604
226   P Simran Singh       31   499    350    20    63   23   174  142.571429  24.950000   78.156313  49.714286
237          PP Shaw       79  1347    946    51   193   37   416  142.389006  26.411765   73.793615  43.974630
355        YK Pathan       33   361    256    12    47   16   125  141.015625  30.083333   78.670360  48.828125
98           E Lewis       22   420    298    13    49   21   145  140.939597  32.307692   76.666667  48.657718
65           CA Lynn       40   779    553    17    99   37   271  140.867993  45.823529   79.332478  49.005425
338         V Sehwag      103  1593   1160    60   230   46   547  137.327586  26.550000   75.078468  47.155172
280         SA Yadav       64   859    628    17   120   27   296  136.783439  50.529412   74.738068  47.133758
250      RA Tripathi       70   977    716    34   123   36   326  136.452514  28.735294   72.466735  45.530726
108       GJ Maxwell       46   433    318    20    59   16   155  136.163522  21.650000   76.674365  48.742138
126       JC Buttler       87  1613   1189    39   207   65   580  135.660219  41.358974   75.511469  48.780488

1.Travis Head have a good strikerate almost 200 and 50+ average also 83.5 boundary% .

2.Sunil Narine and Abishek Sharma good at striking the ball at the powerplay .

3.Players like Yashasvi Jaiswal,Jhonny Bairstow ,P Simran singh ,Prithvi Shaw have a strikerate 140+ is ready to hit in the powerplay.